### ISSUE TITLE
When connecting using a modem, nothing is displayed

### ISSUE THREAD

**Issue by rdelaplante** [klamonte/qodem#61]
When qodem connects to a BBS using a dial-up modem, none of the received bytes are displayed.  

When qodem connects to the same BBS using telnet instead of a dial-up modem then everything comes through fine.

When I dial the same BBS using minicom, everything comes through fine.  I've also tried dialing 56K ISPs as well to see if I can get their basic login prompt.  The login prompt displays and can be interacted with using minicom, but I get a blank screen with qodem.

I tested using a US Robotics Sportster 56K external modem, a US Robotics Courier 56K external modem, and a MultiTech 56K PCI Express internal modem on Ubuntu Linux 17.04.  

I have the exact same issue with syncterm. I think there is a bug in qodem and syncterm because I don't have an issue using any of the modems with minicom.

**Comment by klamonte** [comment]
Thank you for this report.

Are you able to communicate with the modem itself without dialing?  Can you Alt-9, then "AT", and see "OK"?  If not, how does it work with and without hardware flow control (Alt-Y, "w" to toggle CTS/RTS)?

**Comment by rdelaplante** [comment]
Sorry for the late reply.  When I use ALT-9 to toggle serial on/off, I'm not able to see the AT characters I type, or an OK response after I press enter. I've tried it with serial on and off.  I also tried the Alt-Y "w" to toggle CTS/RTS on/off and still am not able to see myself type AT<enter> or an OK response.  I've tried all combinations of these without success.  There doesn't seem to be a local echo on/off option in qodem.

When I choose an item from the phone book it dials and connects successfully.  I just get nothing after it connects.

In syncterm, I am able to go into serial mode and interact with the modem.  I can type AT and see my characters echo'd back to me by the modem, and the OK response. I can ATDT a number and have it dial, but I get nothing after it connects.  The developer suspects something related to DCD and DSR:

https://sourceforge.net/p/syncterm/feature-requests/5/

**Comment by klamonte** [comment]
That is really weird!  I will try to replicate over here and see what I can find out.

**Comment by klamonte** [comment]
I saw something similar today trying to simulate host mode modem answering over a null modem cable.  I am wondering if DCD is not being set by the modem and/or serial driver.  Qodem looks for it and drops the connection really quickly if it isn't set.

Can you try turning on session log (Alt-0), try the dial, and seeing if you see "OFFLINE: modem DCD line went down, lost carrier" in ~/qodem/session_log.txt?  If so, I can add an option to ignore DCD and that might get you back in business.

**Comment by rdelaplante** [comment]
I pressed ALT-0 to enable session logging and it asked me the filename. It had session_log.txt as the default value, so I pressed enter.  Next I pressed ALT-D and scrolled to a local number and pressed enter.  It dialed, and after the long handshake completed (I could hear it in the speaker), qodem immediately crashed with this message: "Segmentation fault (core dumped)".  Below is the entire contents of the log file:

* - * Qodem 1.0alpha Log Generated Tue, 30 May 2017 21:18:42 -0400 BEGIN * - *
 
[2017-05-30 21:19:01] Dialing Teksavvy Dial-Up ISP (4166810302)...

**Comment by klamonte** [comment]
Thanks, that is oddly handy!  :-)  I see where that crash is in phonebook.c (need to change %s to %d on line 8494).  I'll have another fix uploaded probably tonight...

**Comment by rdelaplante** [comment]
I just cloned the git repo and compiled.   How do I enable your new Q_OPTION_MODEM_IGNORE_DCD setting?

**Comment by klamonte** [comment]
Add "modem_ignore_dcd = true" to ~/.qodem/qodemrc and hopefully that will work!

**Comment by rdelaplante** [comment]
There is definitely an improvement with this, although there are some quirks.  

1) When I alt-h to hang up, then alt-d to go to the phone book and pick a different BBS to dial, and press enter, it gets stuck trying to do an AT or ATH0.  I have to hit escape a few times to get out of the phone book, then I see whatever command it tried got an OK. I press alt-d again to go to the phone book and try dialing again.  This time it is able to do ATDT and dial the number.

2) When I dial the 56k ISP, I see the "CONNECT 41333 v42bis", but nothing more until I press enter. Then I get a username prompt and can interact with it just fine.  Maybe waiting for a character before showing the username prompt is their end.. but I seem to remember a welcome message appearing before the username prompt before.

3) When I find a BBS that actually answers the phone and connect, whatever characters come in are totally garbled. See the attached screenshot.  I was having the same issue with minicom as well and hoped that qodem would resolve the issue because it supports proper CP437 characters.  Every single BBS I was able to connect to displayed these garbage characters and I was not able to see a login prompt.  In the phone book it shows that I have the terminal emulation set to ANSI for these BBSes.

![qodem](https://cloud.githubusercontent.com/assets/13222352/26756442/6107c6a2-4870-11e7-8fb3-ee710a7282a9.png)

BTW, the screenshots you took of qodem for your website are of my BBS back around 2003!  The Dog House. I shut it down in July 1998, then brought it back on telnet around 2003 for a while before shutting it down again.  It's awesome to see screenshots of my board on your website!

**Comment by klamonte** [comment]
I thought that might be you :)

I will dig more tomorrow, but at first glance it is looking like Qodem may need to lock dte. There is an option for that, I think it is on the alt-o modem cfg but might be in qodemrc instead.

Sorry for the all the iteration, I don't have access to a land line anymore. But thanks so much for helping fix things :)

**Comment by rdelaplante** [comment]
Take a look at these screenshots.  In the first we see minicom dialing a local ISP.  As soon as it connects it displays some welcome text and a "username:" prompt.  When I press enter it repeats the "username:" prompt.

![minicom-isp](https://cloud.githubusercontent.com/assets/13222352/26766252/c00915ac-495b-11e7-9b64-2c5962b5ec58.png)

When qodem connects it does not display the welcome text or "username:" prompt.   When I press enter, then the server re-sends the "username:" prompt and I finally see it and can interact with it.

![qodem-isp](https://cloud.githubusercontent.com/assets/13222352/26766254/c5cb9dd4-495b-11e7-9f34-aa75b99d6c42.png)

Also, any ideas on why I'm seeing garbled characters when dialing a BBS as shown in my previous comment?

**Comment by klamonte** [comment]
> When I alt-h to hang up, then alt-d to go to the phone book and pick a different BBS to dial, and press enter, it gets stuck trying to do an AT or ATH0. I have to hit escape a few times to get out of the phone book, then I see whatever command it tried got an OK. I press alt-d again to go to the phone book and try dialing again. This time it is able to do ATDT and dial the number.

Thanks for this.  I'll add some more defensive code in the dialer to help stay in sync.

> When I dial the 56k ISP, I see the "CONNECT 41333 v42bis", but nothing more until I press enter. Then I get a username prompt and can interact with it just fine. Maybe waiting for a character before showing the username prompt is their end.. but I seem to remember a welcome message appearing before the username prompt before.

I'm not sure what is going on here.  The next set of patches might improve this, I simplified the routing between dialer and console at the end of the connection.  (A branch of code I thought was working was actually never getting called.)

> When I find a BBS that actually answers the phone and connect, whatever characters come in are totally garbled. See the attached screenshot. I was having the same issue with minicom as well and hoped that qodem would resolve the issue because it supports proper CP437 characters. Every single BBS I was able to connect to displayed these garbage characters and I was not able to see a login prompt. In the phone book it shows that I have the terminal emulation set to ANSI for these BBSes.

My guess (because Qodem and minicom are both doing it) is that the modem is changing its baud after connect, rather than using a fixed DTE baud rate.  Qodem by default locks DTE.  This can be changed on the Alt-O modem config screen, option 'A' to switch DTE Baud from locked to "Varies with connection speed".  There might also be a modem init string to tell it to lock DTE, but that tends to be manufacturer specific;  changing the modem init line to just AT&F or ATZ might help too.

More code will be coming this evening...

**Comment by rdelaplante** [comment]
I had the DTE Baud set to "varies with connection speed" already, so unfortunately that did not help with the garbled characters.  I also tried setting the init string to AT &F ^M but that did not help.

**Comment by klamonte** [comment]
I am getting a bit stumped on the garbled characters.  Next option: is the remote side using UTF-8?  You could try using X_UTF8 emulation to see quickly.  We might just need to resort to a raw capture file to see what the actual byte stream looks like.

**Comment by rdelaplante** [comment]
Below is the qodem capture.  I had to dial a few BBSes before I got one that displayed output after the CONNECT line.   I tried with X_UTF8 and ANSI emulation, but got garbled text.

[capture.txt](https://github.com/klamonte/qodem/files/1053319/capture.txt)

Next I tried the same test with minicom.   I was actually able to use the BBS, but some of the characters were not displayed nicely (probably extended ASCII).

[minicom-capture.txt](https://github.com/klamonte/qodem/files/1053321/minicom-capture.txt)

**Comment by klamonte** [comment]
The captures look exactly like Qodem has a baud rate mismatch with the modem.  What is minicom's modem init string that is working OK?  Does Qodem work OK with that?

Another option: after connecting, you can hit Alt-Y COM Parameters and start trying other baud rates.  Qodem might be landing on 38400 when the modem needs 57600 or even 230400.  Qodem also maxes at 115200 on modem dials, but I notice USR business class are going to 230400, maybe this is the culprit.

I'm pulling a copy of Ubuntu now to see if the serial port was somehow changed in that kernel version too.

**Comment by rdelaplante** [comment]
It's good we have a diagnosis.  I'll have to try that and answer your modem init string question after work. 

FYI, my two external modems are now in storage far away. I've been using the MultiTech 56K PCI Express internal modem: MT9234ZPX-PCIE.  I just plugged it in and found that it works.  I bought it on eBay for $45 US.

http://www.multitech.com/brands/multimodem-zpx
http://www.multitech.com/documents/publications/data-sheets/86002090.pdf

**Comment by rdelaplante** [comment]
I chose the MultiTech over the US Robotics PCI-E modem because the USR no longer has a speaker.  Also the MultiTech seems to be what server manufacturers provide when you want a modem.  They just work in any OS, and I guess they are reliable.  The model I bought includes voice functionality and is more expensive on most eBay auctions.  The data/fax only model is cheaper.

**Comment by rdelaplante** [comment]
Good news!  I can now get proper characters coming through and have seen proper ANSI on a dial-up modem for the first time since probably 1998!  When I dialed a BBS I saw that it was connected at 38400.  Characters were garbled, and when I pressed a letter it would echo back to me garbled.  I pressed ALT-Y like you suggested and tried 57600. Same issue.  I then tried 115200 and the correct character was echo'd back to me.  I hung up and dialed back in and saw this:

![qodem success](https://user-images.githubusercontent.com/13222352/26858245-1c7df552-4afe-11e7-9bb4-fdec6407da32.png)

Next I went into the modem config and configured the speed to be 115200 and DTE Baud to be locked at 115200.  Now when I dial BBSes, I can see proper characters and ANSI color.  

I might still be having an issue with some where I see nothing at all when connecting.  Either that or their BBS is not working.

**Comment by rdelaplante** [comment]
This screenshot from minicom might be helpful. 

Auto bps detect ... No
Modem has DCD line .. Yes
Status line shows ... DTE speed

![minicom modem settings](https://user-images.githubusercontent.com/13222352/26858370-cfc045de-4afe-11e7-902f-7ad88741227f.png)

**Comment by klamonte** [comment]
Awesome!  Thanks for the screenshots, that helps too.

Are you comfortable if I close this issue?

**Comment by rdelaplante** [comment]
I'm not sure if the DCD problem is Linux only, but I think your defaults are to not ignore DCD.  If I hadn't seen this ticket, I certainly would not know to edit  ~/.qodem/qodemrc and add "modem_ignore_dcd = true" to the end if I want to use a modem.   I don't think there is a way to toggle that from the UI, is there?

Also, should the defaults be to have DTE Baud locked to 115200 if they aren't already?  I don't know if I changed those values while trying to figure out what's wrong.

I don't know how to reference a specific comment above since they aren't numbered.  I wrote a comment 5 days ago that shows a screenshot of minicom dialing an ISP, and qodem.  Minicom showed the welcome screen, qodem did not.  However I was able to press enter in qodem to get a username prompt re-sent.   I suspect that is still an issue?

Thanks for taking the time to get this working!

**Comment by rdelaplante** [comment]
> Are you able to communicate with the modem itself without dialing? Can you Alt-9, then "AT", and see "OK"?

Originally I could not get that to work, but now that the DTE baud is set correctly (and possibly a change you made), I can interact with the modem.  

![modem at command](https://user-images.githubusercontent.com/13222352/26979379-8f993592-4cfc-11e7-80a5-66d2687b9b19.png)

**Comment by klamonte** [comment]
> I'm not sure if the DCD problem is Linux only, but I think your defaults are to not ignore DCD. If I hadn't seen this ticket, I certainly would not know to edit ~/.qodem/qodemrc and add "modem_ignore_dcd = true" to the end if I want to use a modem. I don't think there is a way to toggle that from the UI, is there?

Hmm, that's a good point.  I actually added ignore_dcd for this ticket.  But it really should be in the Alt-O modem config rather than the general options file.  I'll get that switched over.

**Comment by rdelaplante** [comment]
I remember another strange issue when using the USR Courier modem.  It would output:

RINGING
RINGING
RINGING

until the other answer ended.  Normally I think you get "RING", not "RINGING".  I think qodem and/or syncterm didn't recognize "RINGING".  When the modem handshake would complete, the phone book dialer would still be counting down and it still displayed the word "RINGING" instead of the "CONNECT xyz" message.  I'd be stuck on that screen and would have to ALT-H to hang up.

My Courier modem is now in storage far away, so it's not easy for me to re-test that right now.

**Comment by klamonte** [comment]
I moved ignore_dcd out of options and into the modem config (where it should have been in the first place), and hopefully also made the lock dte text a little less ambiguous.  ignore_dcd actually does two things: ignores DCD to determine if still connected, and decides whether or not to use DTR to hang up.  This behavior would permit true 3-wire RS232 to a modem to work smoothly.  Here is what it should look like now:

![modem_screen](https://user-images.githubusercontent.com/4357501/27005369-563a80e0-4deb-11e7-8d45-df2af52bbdbe.png)

This should work the same way as before, can you still dial out with this?

(Also, the changes made earlier should (cross fingers) handle the RINGING case too.)

Thanks very much for your help!

**Comment by klamonte** [comment]
I moved ignore_dcd out of options and into the modem config (where it should have been in the first place), and hopefully also made the lock dte text a little less ambiguous.  ignore_dcd actually does two things: ignores DCD to determine if still connected, and decides whether or not to use DTR to hang up.  This behavior would permit true 3-wire RS232 to a modem to work smoothly.  Here is what it should look like now:

![modem_screen](https://user-images.githubusercontent.com/4357501/27005369-563a80e0-4deb-11e7-8d45-df2af52bbdbe.png)

This should work the same way as before, can you still dial out with this?

(Also, the changes made earlier should (cross fingers) handle the RINGING case too.)

Thanks very much for your help!

**Comment by rdelaplante** [comment]
I pulled and compiled the latest code, removed the new line from ~/.qodem/qodemrc, and used your new Ignore DCD setting in the modem config screen to ignore DCD.   I dialed a BBS and got the login prompt with correct characters and ANSI color first try.  Beautiful.

Next I hung up using ALT-H, waited for it to finish and then went to the phone book.  When I dial it now prefixes extra characters, and it seems to get worse when I try again.  Also something is wrong with the serial port or modem because I it says NO CARRIER, Line busy or modem timed out.  Then it goes into a redial pause, then tries again over and over again.  I tried going ALT-9 and doing ATH0 (which gets OK), but that doesn't help.  I tried exiting the program but that doesn't help.  Same with pulling the phone wire from the wall and plugging it back in.   The next thing I'm going to try is rebooting my computer.

Look at the ATDT line in these screenshots

![aatdt](https://user-images.githubusercontent.com/13222352/27005585-9e852770-4def-11e7-8128-4573ac66b740.png)
![line busy](https://user-images.githubusercontent.com/13222352/27005588-ab000fd8-4def-11e7-81c7-02171c7e61a0.png)
![redial](https://user-images.githubusercontent.com/13222352/27005592-b0ccaea8-4def-11e7-8a78-1590d5ea9305.png)
![redial2](https://user-images.githubusercontent.com/13222352/27005596-bcfac250-4def-11e7-8f8c-3964a626e60b.png)
![dial](https://user-images.githubusercontent.com/13222352/27005597-bf235646-4def-11e7-939b-5032481d5272.png)

**Comment by rdelaplante** [comment]
Ok so same issues after rebooting.  It either says NO CARRIER right away and goes into retry, or it does an ATDT with prefixes such as ATDT647ATDT647-847-2083 which causes the phone company to say "please hang up and try your call again", then I can't dial anymore because it keeps saying NO CARRIER.   I tried with and without the new ignore DCD feature.

**Comment by klamonte** [comment]
Ouch!  I think I see what is happening: the dialer is seeing the OK from the modem init string and getting out of sequence -- it thinks this is the OK from the AT, then it gets the AT's OK for the ATDT and tries the ATDT again which kills the old ATDT.  Can you eliminate the modem init string entirely and see how it behaves?

**Comment by klamonte** [comment]
Might also need to power cycle the modem, or manually do AT&F to restore factory defaults.

**Comment by rdelaplante** [comment]
Ok I removed the modem init string.  I'm stuck in a redial pause loop.  Every time it tries to dial I hear the modem speaker make a click like it picked up the hook, then it says NO CARRIER and I go into a redial loop.   While typing this it tried at least a few times and eventually got to dial.  It dialed with extra characters before the ATDT, so the phone company gave a spoken error.

**Comment by rdelaplante** [comment]
my init string was AT &F^M.   To power cycle the modem I will need to power cycle the computer because it is internal.   I'll reboot now.

**Comment by klamonte** [comment]
Oh sorry! :-(

**Comment by rdelaplante** [comment]
Good news!  After powering off the computer and booting back up again, I tried qodem with no modem init string.  I can dial various BBSes, connect and see their welcome screen, hang up, dial another, and repeat many times without issue.  Now it's feeling like a terminal!

**Comment by klamonte** [comment]
Sweet!  I removed the "defensive" ATDT resend, can you try one more time (sorry!)?  If we get multiple dialouts still working, then I will finally be comfortable on pulling the 1.0.0 release.

**Comment by rdelaplante** [comment]
I put my "AT &F^M" init string back in and began to dial.   I was able to connect and hang up a few times, but there were some strange ATDT issues still.  The first dial was AATDT4166810302.   The second dial was normal.  The third dial just said OK and started counting down from 120 seconds. The fourth dial was ATDT4166AT   I think the fifth was the same, and the sixth finally worked and connected.

**Comment by klamonte** [comment]
Hmm, strange.  So right now, it sounds like with no init string, dialouts with current git head are working mostly reliably, but with init string, dialouts work but only sporadically.  Am I understanding right?

**Comment by rdelaplante** [comment]
Yes I did a git pull and recompiled before writing my last comment.  With init strings there are still some strange issues when dialing.

**Comment by rdelaplante** [comment]
After my last comment I removed the init string, powered off the computer, booted back up and dialed the same number a few times with no issues.   So the issues only happen when there is an init string.

**Comment by klamonte** [comment]
Thank you very much for all this troubleshooting!

I would like to eliminate the init string then by default, and then track a separate issue for dialing out with an init string set.  That might turn into a larger help topic (web or F1 help) later on that can track various models of modems, or find some other way to ensure qodem stays in sync.

**Comment by rdelaplante** [comment]
Sure, ok.  At least it's usable for me now once configured properly.

I was thinking about creating a couple of other tickets.  What do you think about these?
1) Ability to go full screen pressing ALT-ENTER doesn't work on Ubuntu Linux/Unity.  Is it supposed to? I'd to have a real DOS experience full screen in 80x25.
2) The escape key doesn't work for me.  For some reason I have to use `.  I'm accessing it locally on my desktop.

**Comment by rdelaplante** [comment]
Regarding full screen, ignore that.  I discovered that pressing F11 goes full screen, although much larger than 80x25.  If I use the gnome terminal menu bar to zoom in a bunch of times and then go full screen, it looks close enough to 80x25.

**Comment by klamonte** [comment]
Alas Alt-Enter won't ever work for qodem-x11 (the PDcurses X11 backend would need non-trivial rework), but it could work for 'qodem' running under another terminal (e.g. xterm).

On the escape key, we can track a new issue for that.  If you are running in gnome-terminal, you might check to see if ESCDELAY is set to anything, that would cause a long lag between ESC and ncurses reporting it.  But I will also get that ubuntu iso in a VM and see what modern gnome-terminal is doing by default.

**Comment by klamonte** [comment]
I'm going to have to be offline until Monday evening, but that you VERY MUCH again for the ticket report and all your help with getting it working better!

**Comment by rdelaplante** [comment]
You're welcome!  Thank you for writing qodem and maintaining it.   The syncterm devs haven't done anything with my ticket yet.  I don't know for sure, but I think qodem is the only terminal available suitable for BBSing with a modem on Linux now that it is fixed.

**Comment by klamonte** [comment]
I think I found those missing bytes.  I've removed the call to flush_serial_port() (it was a holdover from some old embedded stuff), so in theory qodem should no longer potentially be trashing the first few bytes from a new dialup connection.

Can you confirm that dialing out is still working?  I don't think this will fix the init string, but it should hopefully not break anything else...

**Comment by rdelaplante** [comment]
I did a git pull, compiled and re-tested. It works great!  I dialed a bunch of boards.. the ones that answered showed their initial text immediately.  No blank screens.    I did not have an init string configured.

One strange thing I noticed is that when I log out of a BBS, I see "NO CARRIER" but the counter in the status bar continues counting up.  I use ALT-H to hang up.   I have the "Ignore DCD" setting set to "Carrier detect ignored, hangup with string".  Maybe that's to be expected in this mode.   It's probably not safe to look for the string "NO CARRIER" because that could show up in a BBS message area, etc.

**Comment by rdelaplante** [comment]
Here's something interesting.. out of curiosity I changed the "Ignore DCD" setting to "Carrier detect honored, hangup with DTR".  It works perfectly.  I can dial numbers, connect, see their welcome screen, interact, and when they hang up on me qodem stops counting up in the status bar.  I'm able to go to the phone book and dial another number.  So, it looks like the original problem we were trying to solve wasn't related to DCD in the first place.  Or at least DCD issues were just a symptom.    

What advice do you have for the syncterm devs to help them diagnose their issue?  Dial a BBS, no bytes are shown.  Blank screen and I have to hang up.

**Comment by klamonte** [comment]
I'm not sure what would help SyncTERM most.  Ignoring DCD is a good option to have no matter what (and I should have had already but forgot).  But I also added calls to tcdrain() during the dialing sequence and hangup sequence which likely helped a lot: turns out Linux will close the port even if there is data in the UART buffer and never send that.  Finally, I had some awful code after connect that was consuming data (flush_serial_port()) that should never have been there in the first place, and is probably 60% of the real problem.

Thank you so much for your help in debugging.  I just released 1.0.0 a moment ago.  :)

Soon will be on to 1.0.1, but first a break...
